3rd task of testing

IvayloTsvetkov opened the pull request:
Warehouse class with add, sell and sort methods tested in JUnit

mgenov commented on testing_3/test/com/company/WarehouseTest.java:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  @Test
+  public void add() {
```
Test behaviour, not method.

mgenov commented on testing_3/src/com/company/Warehouse.java:
```diff
@@ -0,0 +1,47 @@
+    return products.add(product);
+  }
+
+  public boolean addAmount(String string, int quantity) {
```
String string  doesn't reads good. String productName sounds clear that product name is provided.

mgenov commented on testing_3/test/com/company/WarehouseTest.java:
```diff
@@ -0,0 +1,61 @@
+
+  @Test
+  public void sort() {
+    Product p1 = new Product("Whisky", 16.4);
```
Try to shorten this initialization.

mgenov commented on testing_3/src/com/company/Warehouse.java:
```diff
@@ -0,0 +1,47 @@
+    return products;
+  }
+
+  public List<Product> sortByPrice() {
```
what happens when you call: warehouse.sortByPrice()  and then: warehouse.getProducts() ?

will the products be sorted or not ?

IvayloTsvetkov replied:
yes, they'll be sorted.

mgenov replied:
so If I call getX() it will return X but if I call getX() getY() getX() it will return X'. This doesn't sound reasonable.

mgenov commented on testing_3/src/com/company/Product.java:
```diff
@@ -0,0 +1,54 @@
+    this.name = name;
+    this.price = price;
+    this.quantity = quantity;
+    this.maxQuantity = maxQuantity;
```
This should be part of the warehouse, not of the product. This means that when I buy product I need to have a maxQuantity property.

IvayloTsvetkov commented:
I'm not very proud with my solution, but here it is. Might shorten it once i read about guava.

mgenov commented on testing_3/test/com/company/WarehouseTest.java:
```diff
@@ -0,0 +1,65 @@
+  @Before
+  public void create() {
+    warehouse = new Warehouse();
+    list = new ArrayList<>();
```
```java
List<Product> products = Arrays.asList(
      new Product("Whisky", 16.7d), 
      new Product("Menta", 7.4)
); 
```

mgenov commented on testing_3/src/com/company/Warehouse.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.List;
+
+public class Warehouse {
+  private List<Inventory> inventories = new LinkedList<>();
```
Just inventory should be enough as 1 warehouse has 1 inventory.

mgenov commented on testing_3/src/com/company/Warehouse.java:
```diff
@@ -0,0 +1,68 @@
+public class Warehouse {
+  private List<Inventory> inventories = new LinkedList<>();
+
+  public boolean add(Inventory inventory) {
```
As Inventory will become single item this method becomes obsolete.

mgenov commented on testing_3/src/com/company/Warehouse.java:
```diff
@@ -0,0 +1,68 @@
+    return temp;
+  }
+
+  public void sortByPrice() {
```
This is an example expectation: 
```java
assertThat(warehouse.getProductsSortedByPrice(), is(equalTo(Arrays.asList(new Product(x), new Product(y)));
```

mgenov commented on testing_3/src/com/company/Warehouse.java:
```diff
@@ -0,0 +1,68 @@
+    }
+  }
+
+  public String toString() {
```
Use equals for object comparison. 

kotlin data classes are generating them for you or you can generate them in java from IDEA.

mgenov commented on testing_3/src/com/company/Inventory.java:
```diff
@@ -0,0 +1,53 @@
+
+public class Inventory {
+  private List<Product> products = new ArrayList<>();
+  private int maxQuantity;
```
MaxQuantity is per product item, not per inventory. 

class Inventory -> List of inventory slots, List of products in warehouse
class InventorySlot { name, maxQuantity }


class Product {name, price, quantity }


 Order (List<ProductItem> name, quantity) 
SellResult sell(Order order) 

```java
warehouse.recordProduct("apples", 20, 1d)
SellResult result = warehouse.sell(new Order(new ProductItem("apples", 5))); 
assertThat(result.amount, is(equal(5d))
```

IvayloTsvetkov commented:
Fixed.

IvayloTsvetkov commented:
Fixed.

mgenov commented:
Which one ?

IvayloTsvetkov commented:
The structure. Warehouse now contain Inventory and Inventory contains list of InventorySlots. I also shorten the initialization by using Arrays.asList as suggested.

mgenov commented on testing_3/test/com/company/WarehouseTest.java:
```diff
@@ -0,0 +1,53 @@
+            new InventorySlot(new Product("Vodka", 14.6), 20),
+            new InventorySlot(new Product("Whisky", 16.9), 10)
+    )));
+    Warehouse warehouse2 = new Warehouse(inventory2);
```
Define both warehouses here as it's hard to read when half of the definitions are in the setup and half of them are in the test.

mgenov commented on testing_3/src/com/company/Warehouse.java:
```diff
@@ -0,0 +1,48 @@
+    inventory.sortByPrice();
+  }
+
+  public Inventory getSortedByPrice() {
```
This is a smell as caller is able to use the inventory method directly.

IvayloTsvetkov replied:
I don't like it either, but I don't know how to fix it. I was wondering if I can use comparator directly on Warehouse to sort, but I didn't find a way to do it.

mgenov replied:
The problem is that in the InvetorySlot you are keeping the current quantity and the price. If they are closer to the Warehouse then the problem should be solved.

IvayloTsvetkov commented:
I fixed the sort so there is no more sortByPrice() in the Inventory and removed the Inventory parameter in the getProductSlot(). Created isHavingEnough() and isHavingSpace() in InventorySlot so I can make sell() and record() in Warehouse more readable. Made initialization in of Warehouse in each test and removed the @Before.

mgenov commented on testing_3/src/com/company/Warehouse.java:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  public boolean record(String productName, int quantity) {
+    InventorySlot temp = inventory.findProductSlot(productName);
```
s/temp/inventorySlot/g

IvayloTsvetkov replied:
I'm not sure what is this suppose to mean.

mgenov replied:
replace value temp with inventorySlot

mgenov commented on testing_3/test/com/company/WarehouseTest.java:
```diff
@@ -0,0 +1,52 @@
+    Warehouse warehouse2 = new Warehouse(new Inventory(new ArrayList<InventorySlot>(Arrays.asList(
+            new InventorySlot(new Product("Vodka", 14.6), 20),
+            new InventorySlot(new Product("Whisky", 16.9), 10)))));
+    assertThat(warehouse.getSortedByPrice().toString(), equalTo(warehouse2.toString()));
```
Relaying on toString for comparison is not a good idea.  Compare values not strings as everyone is able at any time to override toString and this test is going to break.

IvayloTsvetkov replied:
![screenshot from 2017-08-15 09-45-52](https://user-images.githubusercontent.com/29853795/29304957-b9682154-819e-11e7-8e69-426435d2c13d.png)

IvayloTsvetkov replied:
this is what i get when i removed the toString().

mgenov replied:
Now I got that you don't know what equals method of objects does. 

This is from the tutorial: 
https://docs.oracle.com/javase/tutorial/java/IandI/objectclass.html

IvayloTsvetkov commented:
Renamed temp to inventorySlot and fixed sortByPrice() test to use equals() instead of toString().
